src/common/runtime/helper/test_worker.ts

```diff
@@ -2,95 +2,125 @@ import { LogMessageWithStack } from '../../internal/logging/log_message.js';
+    const request: WorkerTestRunRequest = { rec, query, expectations };
+    await this.makeRequestAndRecordResult(this.worker, request);
   }
 }
 
 export class TestWorker extends TestDedicatedWorker {}
 
-export class TestSharedWorker {
-  private readonly ctsOptions: CTSOptions;
+export class TestSharedWorker extends TestBaseWorker {
   private readonly port: MessagePort;
-  private readonly resolvers = new Map<string, (result: LiveTestCaseResult) => void>();
 
   constructor(ctsOptions?: CTSOptions) {
-    this.ctsOptions = { ...(ctsOptions || kDefaultCTSOptions), ...{ worker: 'shared' } };
+    super('shared', ctsOptions);
     const selfPath = import.meta.url;
     const selfPathDir = selfPath.substring(0, selfPath.lastIndexOf('/'));
     const workerPath = selfPathDir + '/test_worker-worker.js';
     const worker = new SharedWorker(workerPath, { type: 'module' });
     this.port = worker.port;
     this.port.start();
-    this.port.onmessage = ev => {
-      const query: string = ev.data.query;
-      const result: TransferredTestCaseResult = ev.data.result;
-      if (result.logs) {
-        for (const l of result.logs) {
-          Object.setPrototypeOf(l, LogMessageWithStack.prototype);
-        }
-      }
-      this.resolvers.get(query)!(result as LiveTestCaseResult);
+    this.port.onmessage = ev => this.onmessage(ev);
+  }
+
+  async run(
+    rec: TestCaseRecorder,
+    query: string,
+    expectations: TestQueryWithExpectation[] = []
+  ): Promise<void> {
+    const request: WorkerTestRunRequest = { rec, query, expectations };
+    await this.makeRequestAndRecordResult(this.port, request);
+  }
+}
 
-      // MAINTENANCE_TODO(kainino0x): update the Logger with this result (or don't have a logger and
-      // update the entire results JSON somehow at some point).
-    };
+export class TestServiceWorker extends TestBaseWorker {
+  constructor(ctsOptions?: CTSOptions) {
+    super('service', ctsOptions);
   }
 
   async run(
     rec: TestCaseRecorder,
     query: string,
     expectations: TestQueryWithExpectation[] = []
   ): Promise<void> {
-    this.port.postMessage({
-      query,
-      expectations,
-      ctsOptions: this.ctsOptions,
-    });
-    const workerResult = await new Promise<LiveTestCaseResult>(resolve => {
-      this.resolvers.set(query, resolve);
+    const [suite, name] = query.split(':', 2);
+    const fileName = name.split(',').join('/');
+    const serviceWorkerPath = `/out/${suite}/webworker/${fileName}.worker.js`;
+
+    const registration = await navigator.serviceWorker.register(serviceWorkerPath, {
+      type: 'module',
     });
-    rec.injectResult(workerResult);
+    await registration.update();
+    navigator.serviceWorker.onmessage = ev => this.onmessage(ev);
+    assert(!!registration.active);
+    const request: WorkerTestRunRequest = { rec, query, expectations };
+    await this.makeRequestAndRecordResult(registration.active, request);
+    void registration.unregister();
```

Comment:
Running this locally on /standalone/ it's very slow. I think the problem is that we're creating and destroying a ServiceWorker for every test case. We really only want to create one ServiceWorker for each .worker.js, at most (= one per spec file). Which means we need to unregister later than this anyway.

This can be a followup though.